### DISCUSSION TITLE
Scrolloff

### DISCUSSION THREAD

**Discussion by thinusbaun** [saulpw/visidata#993] [Ideas]
Hi!
First of all, thank you very much for this wonderful piece of software!

I am using VisiData for analyzing logs from piece of software where surroundings of particular line are important. 'zz' keybind does it's job but it would be nice to have scrolling done automatically. In vim there is 'scrolloff' option which determines minimal amount of lines to keep above and below the cursor. When it is set to big number, for example 999, it keeps line with cursor in the middle of screen. It would be nice to have similar option in VisiData. 

I think I would be able to implement this feature, but it would be nice to hear some tips about implementation and API.

**Comment by geekscrapy** [comment]
I'm unsure if this fits what you want exactly, but it works in certain cases for me (warning, my rubbish coding 😏): https://github.com/geekscrapy/visidata_plugins/blob/master/scroll_fix.py

This will only work with cursor up and down, not paging etc. My aim was to keep the cursor in the middle after operations such as search etc. but they don't use cursor up down etc. (of course).

I have discussed with @saulpw that it'd be nice to get a hook into the current selected row, but I've not been able to implement/test some of those suggestions.

So, for me, this works ok, but I'd love for it to be universal, and be able to hook into changes to the current row. I have a couple of usecases for this hook now

**Reply by thinusbaun** [reply]
I've pushed initial implementation of configurable scrolloff [here](https://github.com/thinusbaun/visidata/commit/f11b2c8799d5103167967a25388e18772bf8dddd).
As far as I tested I see some issues:

- [ ] Somehow handle bindings like PG_UP/PG_DOWN where cursor is moved and then topRowIndex/bottomRowIndex is changed. Hook that You have mentioned seems like nice solution.
- [ ] First move within sheet after changing scrolloff value can cause cursor to jump, which doesn't look nice

I haven't tested it with multiline rows, so I don't know how it will work with them.

**Reply by geekscrapy** [reply]
Nice!
I've not tested, but with these modifications are you able to keep the cursor on, for example, the 10th row?

My implementation works so-so with multiline, but there is a related issue for that #916

**Reply by thinusbaun** [reply]
Unfortunately this doesn't work with multiline.
